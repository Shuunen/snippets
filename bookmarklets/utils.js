/* eslint-disable-next-line no-unused-vars */
class Shuutils {
  constructor (app) {
    app = app || {
      id: 'shu-app',
      debug: true
    }
    this.app = app
    this.accentsIn = 'ÀÁÂÃÄÅĄàáâãäåąßÒÓÔÕÕÖØÓòóôõöøóÈÉÊËĘèéêëęðÇĆçćÐÌÍÎÏìíîïÙÚÛÜùúûüÑŃñńŠŚšśŸÿýŽŻŹžżź'
    this.accentsOut = 'AAAAAAAaaaaaaaBOOOOOOOOoooooooEEEEEeeeeeeCCccDIIIIiiiiUUUUuuuuNNnnSSssYyyZZZzzz'
  }

  log (...stuff) {
    if (!this.app.debug) {
      return
    }
    stuff.unshift(this.app.id + ' :')
    console.log.apply(console, stuff)
  }

  warn (...stuff) {
    stuff.unshift(this.app.id + ' :')
    console.warn.apply(console, stuff)
  }

  error (...stuff) {
    stuff.unshift(this.app.id + ' :')
    console.error.apply(console, stuff)
  }

  readableString (str) {
    return str.split('').map(function (letter) {
      var i = this.accentsIn.indexOf(letter)
      return i !== -1 ? this.accentsOut[i] : letter
    }).join('').replace(/(\W|-)/gi, ' ').replace(/\s+/, ' ')
  }

  debounce (func, wait, immediate) {
    var timeout
    return function () {
      var context = this
      var args = arguments
      var later = function later () {
        timeout = null
        if (!immediate) {
          func.apply(context, args)
        }
      }
      var callNow = immediate && !timeout
      clearTimeout(timeout)
      timeout = setTimeout(later, wait)
      if (callNow) {
        func.apply(context, args)
      }
    }
  }

  findOne (selector, context, dontYell) {
    context = context || document
    var item = context.querySelector(selector)
    if (item && this.app.debug) {
      this.log('found element matching "' + selector + '"')
    } else if (!item && !dontYell) {
      this.warn('found no element for selector "' + selector + '"')
    }
    return item
  }

  findFirst (selector, context, dontYell) {
    return this.findAll(selector, context, dontYell)[0]
  }

  findAll (selector, context, dontYell) {
    if (!selector || !selector.length || selector.length === 1) {
      this.error('incorrect selector : ', selector)
    }
    context = context || document
    var items = Array.prototype.slice.call(context.querySelectorAll(selector))
    if (items.length && this.app.debug) {
      this.log('found', items.length, 'elements matching "' + selector + '"')
    } else if (!items.length && !dontYell) {
      this.warn('found no elements for selector "' + selector + '"')
    }
    return items
  }
}
