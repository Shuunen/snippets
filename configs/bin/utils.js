import { copyFile, mkdir } from 'node:fs/promises'
import path from 'node:path'

/**
 * Get the filename from a filepath
 * @param {string} filepath the filepath to get the filename from
 * @returns {string} the filename
 * @example filename('C:\\Users\\me\\file.txt') // 'file.txt'
 */
export function filename(filepath = '') {
  return /[/\\](?<name>[\w.-]+)$/u.exec(filepath)?.groups?.name ?? ''
}

/**
 * Remove lines matching a regexp list
 * @param {string} content the content to clean
 * @param {RegExp[]} regexList the regex list to match
 * @returns {string} the cleaned content
 */
export function removeLinesMatching(content, regexList) {
  const lines = content.split('\n')
  const filteredLines = lines.filter(line => !regexList.some(regex => regex.test(line)))
  return filteredLines.join('\n').trim()
}

/**
 * Remove lines after a given regex
 * @param {string} content the content to remove lines from
 * @param {RegExp} regex the regexp to match the line after which to remove content
 * @returns {string} the content without the lines after the matching line
 */
export function removeLinesAfter(content, regex) {
  const lines = content.split('\n')
  const index = lines.findIndex(line => regex.test(line))
  if (index === -1) return content.trim()
  return lines.slice(0, index).join('\n').trim()
}

/**
 * Convert carriage return to unix line endings
 * @param {string} content the content to be processed
 * @returns the processed content with unix line endings
 */
export function useUnixCarriageReturn(content) {
  return content.replace(/\r\n/gu, '\n')
}

/**
 * Clean a file details content
 * @param {string} content the file content to clean
 * @param {RegExp} [linesAfter] a regex to remove lines after
 * @param {RegExp[]} [linesMatching] a list of regex to remove lines matching
 * @param shouldClearSpaces if true will also clear spaces
 * @returns {string} the cleaned file content
 */
// eslint-disable-next-line @typescript-eslint/max-params
export function clean(content, linesAfter, linesMatching, shouldClearSpaces = true) {
  if (!content) return ''
  let output = content
  if (linesAfter) output = removeLinesAfter(output, linesAfter)
  if (linesMatching) output = removeLinesMatching(output, linesMatching)
  if (shouldClearSpaces) output = output.replace(/\s*/gu, '')
  return output
}

/* c8 ignore start */

/**
 * Normalize a filepath with slash style
 * @param {string} filepath the filepath
 * @param {boolean} shouldUseTilde use tilde will replace the home directory with ~
 * @param {string} home the home directory path
 * @returns the normalized path
 */
export function normalizePathWithSlash(filepath, shouldUseTilde = false, home = process.env.HOME ?? '') {
  let outPath = path.normalize(filepath).replace(/\\/gu, '/')
  if (shouldUseTilde) outPath = outPath.replace(normalizePathWithSlash(home), '~')
  return outPath
}

/**
 * Copy a file
 * @param {string} source the source file
 * @param {string} destination the destination file
 * @returns {Promise<boolean>} some bool result; i dont know im in the train to Paris
 */
export async function copy(source, destination) {
  // destination will be created or overwritten by default.
  const destinationFolder = destination.replace(filename(destination), '')
  // eslint-disable-next-line @typescript-eslint/naming-convention
  await mkdir(destinationFolder, { recursive: true })

  return copyFile(source, destination)
    .then(() => true)
    .catch(error => {
      console.log(error)
      return false
    })
}
/* c8 ignore stop */
