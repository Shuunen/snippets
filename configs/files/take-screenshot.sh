#!/bin/bash

time=$(zenity --entry --title="Take screenshot" --text="Please type the time in mmss or ss :" --entry-text="")

node "$HOME/Projects/github/snippets/one-file/take-screenshot.js" "$1" "$time"
