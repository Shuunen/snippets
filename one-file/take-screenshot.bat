REM set need to be here, outside the logging section

REM the video file we want to set title to
REM like : "D:\Downloads\Big Buck Bunny (2003)\Big Buck Bunny (2003).mkv"
set targetFile=%1

REM this script folder
REM like : "D:\Apps\Set.Title\"
SET thisFolder=%~dp0

REM DO NOT try to pipe the output of this script to a file
REM because we need the user input in the console
ts-node-esm %thisFolder%/take-screenshot.cli.ts %targetFile%
