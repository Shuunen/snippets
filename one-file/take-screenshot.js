#!/usr/bin/env node
import { exec } from 'child_process'
import { promises as fs } from 'fs'
import path from 'path'
import { createInterface } from 'readline'
import { fileURLToPath } from 'url'

const ask = createInterface({ input: process.stdin, output: process.stdout })

const shellCommand = async cmd => new Promise(resolve => {
  exec(cmd, (error, stdout, stderr) => {
    if (error) console.error(error)
    resolve(stdout || stderr)
  })
})

const deleteFile = path => fs.unlink(path).catch(() => 'ok')

const getVideoMetadata = async path => {
  const output = await shellCommand(`ffprobe -show_format -show_streams -print_format json -v quiet -i "${path}" `)
  if (output[0] !== '{') throw new Error('ffprobe output should be JSON but got :' + output)
  const data = JSON.parse(output)
  const video = data.streams.find(stream => stream.codec_type === 'video')
  const height = video.height
  const media = data.format || {}
  const duration = media.duration
  const title = (media.tags && media.tags.title) || ''
  const size = (media.size !== undefined && (media.size > 0 || media.size.length > 0)) ? media.size : getFileSize(path)
  return { title, size, height, duration }
}

const getFileSize = async path => (await fs.stat(path)).size

const readableSize = size => {
  let unit = 'go'
  let nb = (size / 1024 / 1024 / 1024).toFixed(1)
  if (nb[0] === '0') {
    unit = 'mo'
    nb = String(Math.round(size / 1024 / 1024))
  }
  return nb.replace('.', ',') + unit
}

const readableDuration = seconds => new Date(seconds * 1000).toISOString().slice(11, 19).replace(':', 'h').replace(':', 'm') + 's'

const readableHeight = height => height ? `${height}p` : ''

const currentFolder = path.dirname(fileURLToPath(import.meta.url))
const logFile = path.join(currentFolder, 'take-screenshot.log')
const modulos = [5, 4, 3, 2, 1]
const logClear = () => fs.writeFile(logFile, '')
const logAdd = (...stuff) => fs.appendFile(logFile, stuff.join(' ') + '\n')

const asciiWelcome = () => {
  console.log(`
  ~|~ _ |  _   (~ _ _ _  _  _  _|_  _ _|_
   | (_||<(/_  _)(_| (/_(/_| |_\\| |(_) |
  `)
}

const getTask = (totalSeconds, title, videoPath, meta = {}) => {
  const minutes = Math.floor(totalSeconds / 60)
  const seconds = totalSeconds % 60
  const timeHuman = `${(minutes > 0 ? `${minutes}m` : '') + seconds}s`.trim()
  const screenName = `${title.replace(/\./g, ' ')} ${timeHuman} ${readableSize(meta.size)} ${readableHeight(meta.height)} ${readableDuration(meta.duration)}.jpg`.replace(/\s?["*/:<>?\\|]+\s?/g, ' ').replace(/\s+/g, ' ') // replace un-authorized characters in filename
  const screenPath = path.join(process.env.HOME || process.env.USERPROFILE, 'Pictures', screenName)
  return { totalSeconds, videoPath, screenPath }
}

const getTasks = async input => {
  const { a, b, modulo = '' } = String(input).match(/(?<a>\d{1,2})(?<b>\d{1,2})?(?<modulo>[+-]{1,2})?/).groups
  const seconds = Number.parseInt(b || a, 10)
  const minutes = b === undefined ? 0 : Number.parseInt(a, 10)
  const videoPath = process.argv[2]
  const videoName = path.basename(videoPath)
  const meta = await getVideoMetadata(videoPath)
  const title = meta.title.length > videoName.length ? meta.title : videoName
  await logAdd(`- minutes : ${minutes}\n- seconds : ${seconds}\n- modulo : ${modulo}\n- videoPath : ${videoPath}\n- videoName : ${videoName}\n- title : ${title}`)
  const tasks = []
  const totalSeconds = (minutes * 60) + seconds
  if (modulo.includes('-')) modulos.forEach(m => tasks.push(totalSeconds - m))
  tasks.push(totalSeconds)
  if (modulo.includes('+')) modulos.sort().forEach(m => tasks.push(totalSeconds + m))
  await logAdd(`- tasks (seconds) : ${tasks.toString()}`)
  return tasks.map(tSecs => getTask(tSecs, title, videoPath, meta))
}

const takeScreenAt = async input => {
  const tasks = await getTasks(input).catch(error => console.error(error))
  for (let task of tasks) {
    const { totalSeconds, videoPath, screenPath } = task
    if (!totalSeconds) throw new Error('failed to process variables')
    const cmd = `ffmpeg -ss ${totalSeconds} -i "${videoPath}" -frames:v 1 -q:v 1 "${screenPath}"`
    await deleteFile(screenPath)
    await logAdd('Command :', cmd)
    await logAdd(await shellCommand(cmd))
  }
  process.exit(0)
}

const init = async () => {
  asciiWelcome()
  await logClear()
  await logAdd('Take screenshot starts @', new Date())
  if (!process.argv[2]) throw new Error('missing videoPath')
  if (process.argv[3]) return takeScreenAt(process.argv[3])
  ask.question('  Please type the time in mmss or ss : ', time => takeScreenAt(time))
}

init().catch(error => console.error(error))
