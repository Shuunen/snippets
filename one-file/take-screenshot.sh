#!/bin/sh

ts-node-esm --transpileOnly "$HOME/Projects/github/snippets/one-file/take-screenshot.cli.ts" "$1" "$2"
