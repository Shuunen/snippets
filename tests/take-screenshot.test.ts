import { check, checksRun } from 'shuutils'
import { emptyMetadata, getFfmpegCommand, getScreenshotFilename, parseUserInput, parseVideoMetadata, readableDuration, readableSize } from '../one-file/take-screenshot.utils'

const ffProbeOutputA = { streams: [] }

check('parseVideoMetadata A', parseVideoMetadata(ffProbeOutputA), emptyMetadata)

const ffProbeOutputB = {
  streams: [{
    // eslint-disable-next-line @typescript-eslint/naming-convention, camelcase
    codec_type: 'audio',
    width: 0,
    height: 0,
    duration: '',
  }, {
    // eslint-disable-next-line @typescript-eslint/naming-convention, camelcase
    codec_type: 'video',
    width: 1920,
    height: 1080,
    duration: '00:00:00.000000',
  }],
  format: {
    filename: 'plop.mp4',
    tags: {
      title: 'plop and the doe',
    },
    size: 123_456,
    duration: 789,
  },
}

const expectedMetadataB = {
  title: 'plop and the doe',
  size: 123_456,
  height: 1080,
  duration: 789,
}

check('parseVideoMetadata B', parseVideoMetadata(ffProbeOutputB), expectedMetadataB)

const ffProbeOutputCnoVideo = {
  streams: [{
    // eslint-disable-next-line @typescript-eslint/naming-convention, camelcase
    codec_type: 'audio',
    width: 0,
    height: 0,
    duration: '',
  }],
  format: {
    filename: 'jazz.mp3',
    tags: {
      title: 'couch jazz',
    },
    size: 698_765,
    duration: 123,
  },
}

check('parseVideoMetadata C', parseVideoMetadata(ffProbeOutputCnoVideo), emptyMetadata)

const ffProbeOutputDnoFormat = {
  streams: [{
    // eslint-disable-next-line @typescript-eslint/naming-convention, camelcase
    codec_type: 'video',
    width: 1280,
    height: 720,
    duration: '12',
  }],
}

const expectedMetadataD = {
  title: '',
  size: 0,
  height: 720,
  duration: 0,
}

check('parseVideoMetadata D', parseVideoMetadata(ffProbeOutputDnoFormat), expectedMetadataD)

check('parseUserInput A specific ss', parseUserInput('12'), [{ minutes: 0, seconds: 12 }])
check('parseUserInput B ss+-1', parseUserInput('12+-1'), [{ minutes: 0, seconds: 11 }, { minutes: 0, seconds: 12 }, { minutes: 0, seconds: 13 }])
check('parseUserInput C mmss+-2', parseUserInput('2106+-2'), [{ minutes: 21, seconds: 4 }, { minutes: 21, seconds: 5 }, { minutes: 21, seconds: 6 }, { minutes: 21, seconds: 7 }, { minutes: 21, seconds: 8 }])
check('parseUserInput D mmss around minute gaps', parseUserInput('0100-+1'), [{ minutes: 0, seconds: 59 }, { minutes: 1, seconds: 0 }, { minutes: 1, seconds: 1 }])
check('parseUserInput E mmss around minute gaps', parseUserInput('0100-+2'), [{ minutes: 0, seconds: 58 }, { minutes: 0, seconds: 59 }, { minutes: 1, seconds: 0 }, { minutes: 1, seconds: 1 }, { minutes: 1, seconds: 2 }])
check('parseUserInput F mmss around minute gaps', parseUserInput('0100-+3'), [{ minutes: 0, seconds: 57 }, { minutes: 0, seconds: 58 }, { minutes: 0, seconds: 59 }, { minutes: 1, seconds: 0 }, { minutes: 1, seconds: 1 }, { minutes: 1, seconds: 2 }, { minutes: 1, seconds: 3 }])
check('parseUserInput G mmss around minute gaps', parseUserInput('2201-+2'), [{ minutes: 21, seconds: 59 }, { minutes: 22, seconds: 0 }, { minutes: 22, seconds: 1 }, { minutes: 22, seconds: 2 }, { minutes: 22, seconds: 3 }])
check('parseUserInput H mmss around minute gaps', parseUserInput('2259-+2'), [{ minutes: 22, seconds: 57 }, { minutes: 22, seconds: 58 }, { minutes: 22, seconds: 59 }, { minutes: 23, seconds: 0 }, { minutes: 23, seconds: 1 }])
check('parseUserInput I empty input', parseUserInput(''), [])
check('parseUserInput J invalid input', parseUserInput('plop'), [])
check('parseUserInput K invalid input', parseUserInput('12plop'), [])
check('parseUserInput L invalid input', parseUserInput('plop12'), [])
check('parseUserInput M specific mmss', parseUserInput('1234'), [{ minutes: 12, seconds: 34 }])
check('parseUserInput N default to modulo 5 if marker exists', parseUserInput('1234+-'), [{ minutes: 12, seconds: 29 }, { minutes: 12, seconds: 30 }, { minutes: 12, seconds: 31 }, { minutes: 12, seconds: 32 }, { minutes: 12, seconds: 33 }, { minutes: 12, seconds: 34 }, { minutes: 12, seconds: 35 }, { minutes: 12, seconds: 36 }, { minutes: 12, seconds: 37 }, { minutes: 12, seconds: 38 }, { minutes: 12, seconds: 39 }])

check('readableDuration A', readableDuration(0), '00h00m00s')
check('readableDuration B', readableDuration(28), '00h00m28s')
check('readableDuration C', readableDuration(60), '00h01m00s')
check('readableDuration D', readableDuration(61), '00h01m01s')
check('readableDuration E', readableDuration(3600), '01h00m00s')
check('readableDuration F', readableDuration(3612), '01h00m12s')

check('readableSize A', readableSize(0), '0mo')
check('readableSize B', readableSize(1000), '0mo')
check('readableSize C', readableSize(1_000_000), '1mo')
check('readableSize D', readableSize(1_000_000_000), '954mo')
check('readableSize E', readableSize(1_000_000_000_000), '931,3go')
check('readableSize F', readableSize(1_924_654_000), '1,8go')

check('getScreenshotFilename A', getScreenshotFilename(3600, { title: 'plop', size: 123_456_789, height: 1080, duration: 789 }), 'plop 01h00m00s 118mo 1080p 00h13m09s.jpg')
check('getScreenshotFilename B', getScreenshotFilename(126, { title: 'The little John and the Doe are awaiting the arrival of the big John (2013) Encoded by HurrayXXX_36 x265 10bit HDR DTS-HD-MA', size: 1_000_000_000, height: 720, duration: 1_000_000 }), 'The little John and the Doe are awaiting the arrival of the big John (2013) Encoded by HurrayXXX_36 x265 10bit HDR DTS-HD-MA 00h02m06s 954mo 720p 13h46m40s.jpg')
check('getScreenshotFilename C', getScreenshotFilename(0, { title: '', size: 0, height: 0, duration: 0 }), '00h00m00s 0mo 0p 00h00m00s.jpg')

check('getFfmpegCommand A', getFfmpegCommand({ totalSeconds: 120, videoPath: 'plop.mp4', screenPath: 'plop.jpg' }), 'ffmpeg -ss 120 -i "plop.mp4" -frames:v 1 -q:v 1 "plop.jpg"')

checksRun()
