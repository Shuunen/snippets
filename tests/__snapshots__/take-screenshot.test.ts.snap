// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getTargets A 1`] = `
[
  {
    "minutes": 2,
    "seconds": 25,
  },
  {
    "minutes": 2,
    "seconds": 26,
  },
  {
    "minutes": 2,
    "seconds": 27,
  },
  {
    "minutes": 2,
    "seconds": 28,
  },
  {
    "minutes": 2,
    "seconds": 29,
  },
  {
    "minutes": 2,
    "seconds": 30,
  },
  {
    "minutes": 2,
    "seconds": 31,
  },
  {
    "minutes": 2,
    "seconds": 32,
  },
  {
    "minutes": 2,
    "seconds": 33,
  },
  {
    "minutes": 2,
    "seconds": 34,
  },
  {
    "minutes": 2,
    "seconds": 35,
  },
]
`;
